File: day07/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	re := regexp.MustCompile(`^([a-z ]+) bags contain`)
	re2 := regexp.MustCompile(`([0-9]+) ([a-z ]+) bags?[,.]`)
	re3 := regexp.MustCompile(`no other bags.$`)

	graph := map[string]map[string]int{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		m := re.FindStringSubmatch(line)
		if m == nil {
			log.Fatal("Invalid line format")
		}
		m2 := re2.FindAllStringSubmatch(line, -1)
		m3 := re3.MatchString(line)
		graph[m[1]] = map[string]int{}
		if m3 {
			continue
		}
		if m2 == nil {
			log.Fatal("Invalid line format")
		}
		for _, i := range m2 {
			num, err := strconv.Atoi(i[1])
			if err != nil {
				log.Fatal(err)
			}
			graph[m[1]][i[2]] = num
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	count := 0
	cache := map[string]bool{}
	for k := range graph {
		if hasPathTo(k, "shiny gold", graph, cache) {
			count++
		}
	}

	fmt.Println("Part 1:", count)

	cache2 := map[string]int{}
	fmt.Println("Part 2:", countChildren("shiny gold", graph, cache2)-1)
}

func hasPathTo(a, b string, graph map[string]map[string]int, cache map[string]bool) bool {
	if v, ok := cache[a]; ok {
		return v
	}
	for k := range graph[a] {
		if b == k {
			cache[a] = true
			return true
		}
	}
	for k := range graph[a] {
		if hasPathTo(k, b, graph, cache) {
			cache[a] = true
			return true
		}
	}
	cache[a] = false
	return false
}

func countChildren(a string, graph map[string]map[string]int, cache map[string]int) int {
	if v, ok := cache[a]; ok {
		return v
	}
	count := 1
	for k, v := range graph[a] {
		count += v * countChildren(k, graph, cache)
	}
	cache[a] = count
	return count
}

File: day07/src/main.rs
use regex::Regex;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let re = Regex::new(r"^([a-z ]+) bags contain")?;
    let re2 = Regex::new(r"([0-9]+) ([a-z ]+) bags?[,.]")?;
    let re3 = Regex::new(r"no other bags.$")?;

    let graph = {
        let mut graph = HashMap::new();
        for line in reader.lines() {
            let line = line?;
            let m = re.captures(&line).ok_or("Invalid line format")?;
            let m1 = m.get(1).ok_or("Invalid line format")?.as_str().to_owned();
            let m3 = re3.is_match(&line);
            let mut subgraph = HashMap::new();
            if m3 {
                graph.insert(m1, subgraph);
                continue;
            }
            if !re2.is_match(&line) {
                return Err("Invalid line format".into());
            }
            for i in re2.captures_iter(&line) {
                let num = i
                    .get(1)
                    .ok_or("Invalid line format")?
                    .as_str()
                    .parse::<usize>()?;
                let k = i.get(2).ok_or("Invalid line format")?.as_str().to_owned();
                subgraph.insert(k, num);
            }
            graph.insert(m1, subgraph);
        }
        graph
    };

    let mut count = 0;
    let mut cache = HashMap::new();
    for i in graph.keys() {
        if has_path_to(i, "shiny gold", &graph, &mut cache) {
            count += 1;
        }
    }
    println!("Part 1: {}", count);

    let mut cache2 = HashMap::new();
    let count2 = count_children("shiny gold", &graph, &mut cache2);
    println!("Part 2: {}", count2);

    Ok(())
}

fn has_path_to<'a>(
    a: &'a str,
    b: &str,
    graph: &'a HashMap<String, HashMap<String, usize>>,
    cache: &mut HashMap<&'a str, bool>,
) -> bool {
    if let Some(&v) = cache.get(a) {
        return v;
    }
    let edges = match graph.get(a) {
        Some(edges) => edges,
        None => return false,
    };
    for i in edges.keys() {
        if b == i {
            cache.insert(a, true);
            return true;
        }
    }
    for i in edges.keys() {
        if has_path_to(i, b, graph, cache) {
            cache.insert(a, true);
            return true;
        }
    }
    cache.insert(a, false);
    return false;
}

fn count_children<'a>(
    a: &'a str,
    graph: &'a HashMap<String, HashMap<String, usize>>,
    cache: &mut HashMap<&'a str, usize>,
) -> usize {
    if let Some(&v) = cache.get(a) {
        return v;
    }
    let mut count = 1;
    let edges = match graph.get(a) {
        Some(edges) => edges,
        None => return 1,
    };
    for (k, &v) in edges {
        count += v * count_children(k, graph, cache);
    }
    cache.insert(a, count);
    return count;
}
